File: mini-services/trading-engine/app/smc/fvg.py
"""
SMC Engine - Fair Value Gap (FVG) Detection Module
Mathematical implementation of Price Imbalances

Mathematical Formulas:

1. Bullish FVG (3 candle model):
   if candle1.high < candle3.low:
       fvg = True
       gap_top = candle3.low
       gap_bottom = candle1.high

2. Bearish FVG:
   if candle1.low > candle3.high:
       fvg = True
       gap_top = candle1.low
       gap_bottom = candle3.high

3. FVG Fill Calculation:
   fill_percentage = (price_movement / gap_size) * 100

Author: Trading AI Agent
"""

from dataclasses import dataclass
from typing import List, Optional
from enum import Enum
from datetime import datetime
import numpy as np

from .swing import Candle


class FVGType(Enum):
    """Fair Value Gap Types"""
    BULLISH = "BULLISH"
    BEARISH = "BEARISH"


@dataclass
class FairValueGap:
    """Fair Value Gap"""
    type: FVGType
    gap_top: float
    gap_bottom: float
    gap_size: float
    candle_index: int  # Index of middle candle (impulse)
    filled: bool = False
    filled_at: Optional[datetime] = None
    fill_percentage: float = 0.0
    timestamp: Optional[datetime] = None


class FVGDetector:
    """
    Fair Value Gap Detection using mathematical formulas
    
    FVG represents market imbalance where price moved too quickly.
    Price tends to return to fill these gaps.
    """
    
    def __init__(self, min_gap_percent: float = 0.1):
        """
        Initialize FVG Detector
        
        Args:
            min_gap_percent: Minimum gap size as percentage of price
        """
        self.min_gap_percent = min_gap_percent
    
    def detect_bullish_fvg(self, candles: List[Candle]) -> List[FairValueGap]:
        """
        Detect Bullish FVG
        
        Mathematical Formula (3 candle model):
        Candle1 High < Candle3 Low
        
        This creates an upward gap that price may revisit
        
        Gap Zone:
        - gap_top = candle3.low
        - gap_bottom = candle1.high
        """
        if len(candles) < 3:
            return []
        
        fvgs = []
        
        for i in range(1, len(candles) - 1):
            candle1 = candles[i - 1]
            candle2 = candles[i]      # Impulse candle
            candle3 = candles[i + 1]
            
            # Mathematical condition: candle1.high < candle3.low
            if candle1.high < candle3.low:
                gap_top = candle3.low
                gap_bottom = candle1.high
                gap_size = gap_top - gap_bottom
                
                # Calculate gap percentage
                avg_price = (gap_top + gap_bottom) / 2
                gap_percent = (gap_size / avg_price) * 100
                
                # Check minimum gap size
                if gap_percent >= self.min_gap_percent:
                    # Check fill status
                    filled, fill_pct, filled_at = self._check_fill(
                        candles, i + 1, gap_top, gap_bottom, 'BULLISH'
                    )
                    
                    fvg = FairValueGap(
                        type=FVGType.BULLISH,
                        gap_top=gap_top,
                        gap_bottom=gap_bottom,
                        gap_size=gap_size,
                        candle_index=i,
                        filled=filled,
                        filled_at=filled_at,
                        fill_percentage=fill_pct,
                        timestamp=candle2.timestamp
                    )
                    fvgs.append(fvg)
        
        return fvgs
    
    def detect_bearish_fvg(self, candles: List[Candle]) -> List[FairValueGap]:
        """
        Detect Bearish FVG
        
        Mathematical Formula (3 candle model):
        Candle1 Low > Candle3 High
        
        Gap Zone:
        - gap_top = candle1.low
        - gap_bottom = candle3.high
        """
        if len(candles) < 3:
            return []
        
        fvgs = []
        
        for i in range(1, len(candles) - 1):
            candle1 = candles[i - 1]
            candle2 = candles[i]
            candle3 = candles[i + 1]
            
            # Mathematical condition: candle1.low > candle3.high
            if candle1.low > candle3.high:
                gap_top = candle1.low
                gap_bottom = candle3.high
                gap_size = gap_top - gap_bottom
                
                avg_price = (gap_top + gap_bottom) / 2
                gap_percent = (gap_size / avg_price) * 100
                
                if gap_percent >= self.min_gap_percent:
                    filled, fill_pct, filled_at = self._check_fill(
                        candles, i + 1, gap_top, gap_bottom, 'BEARISH'
                    )
                    
                    fvg = FairValueGap(
                        type=FVGType.BEARISH,
                        gap_top=gap_top,
                        gap_bottom=gap_bottom,
                        gap_size=gap_size,
                        candle_index=i,
                        filled=filled,
                        filled_at=filled_at,
                        fill_percentage=fill_pct,
                        timestamp=candle2.timestamp
                    )
                    fvgs.append(fvg)
        
        return fvgs
    
    def detect_all_fvgs(self, candles: List[Candle]) -> List[FairValueGap]:
        """Detect all Fair Value Gaps"""
        bullish = self.detect_bullish_fvg(candles)
        bearish = self.detect_bearish_fvg(candles)
        
        return sorted(bullish + bearish, key=lambda f: f.candle_index)
    
    def _check_fill(self, candles: List[Candle], start_index: int,
                   gap_top: float, gap_bottom: float, 
                   fvg_type: str) -> tuple:
        """
        Check if FVG has been filled
        
        Mathematical Calculation:
        - For Bullish FVG: filled from top (price drops into gap)
        - For Bearish FVG: filled from bottom (price rises into gap)
        
        fill_percentage = (fill_amount / gap_size) * 100
        """
        gap_size = gap_top - gap_bottom
        max_fill = 0.0
        filled_at = None
        
        for i in range(start_index + 1, len(candles)):
            candle = candles[i]
            
            if fvg_type == 'BULLISH':
                # Filled from above (price drops)
                if candle.low <= gap_top:
                    fill_amount = min(gap_top - candle.low, gap_size)
                    fill_pct = (fill_amount / gap_size) * 100
                    
                    if fill_pct > max_fill:
                        max_fill = fill_pct
                        if fill_pct >= 50:
                            filled_at = candle.timestamp
            else:
                # Bearish FVG - filled from below
                if candle.high >= gap_bottom:
                    fill_amount = min(candle.high - gap_bottom, gap_size)
                    fill_pct = (fill_amount / gap_size) * 100
                    
                    if fill_pct > max_fill:
                        max_fill = fill_pct
                        if fill_pct >= 50:
                            filled_at = candle.timestamp
        
        is_filled = max_fill >= 50
        return is_filled, max_fill, filled_at
    
    def get_unfilled_fvgs(self, fvgs: List[FairValueGap]) -> List[FairValueGap]:
        """Get all unfilled FVGs"""
        return [f for f in fvgs if not f.filled]
    
    def get_fvg_at_price(self, price: float, fvgs: List[FairValueGap]) -> Optional[FairValueGap]:
        """Get FVG that contains a price level"""
        for fvg in fvgs:
            if fvg.gap_bottom <= price <= fvg.gap_top:
                return fvg
        return None
    
    def get_nearest_fvg(self, price: float, fvgs: List[FairValueGap],
                       fvg_type: Optional[FVGType] = None) -> Optional[FairValueGap]:
        """Get nearest FVG to current price"""
        filtered = fvgs
        if fvg_type:
            filtered = [f for f in filtered if f.type == fvg_type]
        
        if not filtered:
            return None
        
        def distance(fvg: FairValueGap) -> float:
            mid = (fvg.gap_top + fvg.gap_bottom) / 2
            return abs(price - mid)
        
        return min(filtered, key=distance)

File: src/lib/trading/smc/fvg.ts
// SMC Engine - Fair Value Gap (FVG) Detection Module
// Detects price imbalances that act as magnets

import { Candle, FairValueGap, FVGType } from '../types';

export interface FVGConfig {
  minGapPercent: number;   // Minimum gap size as percentage
  fillThreshold: number;   // Percentage considered filled
}

const DEFAULT_CONFIG: FVGConfig = {
  minGapPercent: 0.1,  // 0.1% minimum gap
  fillThreshold: 50    // 50% filled threshold
};

/**
 * Detect Bullish FVG
 * Forms when candle1.high < candle3.low (gap up)
 */
export function detectBullishFVG(
  candles: Candle[],
  config: FVGConfig = DEFAULT_CONFIG
): FairValueGap[] {
  const fvgs: FairValueGap[] = [];

  if (candles.length < 3) return fvgs;

  for (let i = 1; i < candles.length - 1; i++) {
    const candle1 = candles[i - 1];
    const candle3 = candles[i + 1];
    const candle2 = candles[i]; // The impulse candle

    // Bullish FVG: candle1 high < candle3 low
    if (candle1.high < candle3.low) {
      const gapTop = candle3.low;
      const gapBottom = candle1.high;
      const gapSize = gapTop - gapBottom;
      const avgPrice = (gapTop + gapBottom) / 2;
      const gapPercent = (gapSize / avgPrice) * 100;

      if (gapPercent >= config.minGapPercent) {
        const { filled, fillPercentage, filledAt } = checkFVGFill(
          candles,
          i + 1,
          gapTop,
          gapBottom,
          'BULLISH'
        );

        fvgs.push({
          symbolId: candle2.symbolId,
          timeframe: candle2.timeframe,
          type: FVGType.BULLISH,
          gapTop: gapTop,
          gapBottom: gapBottom,
          candleIndex: i,
          filled: fillPercentage >= config.fillThreshold,
          filledAt: filledAt,
          fillPercentage: fillPercentage
        });
      }
    }
  }

  return fvgs;
}

/**
 * Detect Bearish FVG
 * Forms when candle1.low > candle3.high (gap down)
 */
export function detectBearishFVG(
  candles: Candle[],
  config: FVGConfig = DEFAULT_CONFIG
): FairValueGap[] {
  const fvgs: FairValueGap[] = [];

  if (candles.length < 3) return fvgs;

  for (let i = 1; i < candles.length - 1; i++) {
    const candle1 = candles[i - 1];
    const candle3 = candles[i + 1];
    const candle2 = candles[i];

    // Bearish FVG: candle1 low > candle3 high
    if (candle1.low > candle3.high) {
      const gapTop = candle1.low;
      const gapBottom = candle3.high;
      const gapSize = gapTop - gapBottom;
      const avgPrice = (gapTop + gapBottom) / 2;
      const gapPercent = (gapSize / avgPrice) * 100;

      if (gapPercent >= config.minGapPercent) {
        const { filled, fillPercentage, filledAt } = checkFVGFill(
          candles,
          i + 1,
          gapTop,
          gapBottom,
          'BEARISH'
        );

        fvgs.push({
          symbolId: candle2.symbolId,
          timeframe: candle2.timeframe,
          type: FVGType.BEARISH,
          gapTop: gapTop,
          gapBottom: gapBottom,
          candleIndex: i,
          filled: fillPercentage >= config.fillThreshold,
          filledAt: filledAt,
          fillPercentage: fillPercentage
        });
      }
    }
  }

  return fvgs;
}

/**
 * Detect all FVGs
 */
export function detectFVG(
  candles: Candle[],
  config: FVGConfig = DEFAULT_CONFIG
): FairValueGap[] {
  const bullishFVGs = detectBullishFVG(candles, config);
  const bearishFVGs = detectBearishFVG(candles, config);

  return [...bullishFVGs, ...bearishFVGs];
}

/**
 * Check if FVG is filled
 */
function checkFVGFill(
  candles: Candle[],
  startIndex: number,
  gapTop: number,
  gapBottom: number,
  type: 'BULLISH' | 'BEARISH'
): { filled: boolean; fillPercentage: number; filledAt?: Date } {
  const gapSize = gapTop - gapBottom;
  let maxFill = 0;
  let filledAt: Date | undefined;

  for (let i = startIndex + 1; i < candles.length; i++) {
    const candle = candles[i];

    if (type === 'BULLISH') {
      // Check how much of the gap was filled from the top
      if (candle.low <= gapTop) {
        const fillAmount = Math.min(gapTop - candle.low, gapSize);
        const fillPercent = (fillAmount / gapSize) * 100;
        
        if (fillPercent > maxFill) {
          maxFill = fillPercent;
          if (fillPercent >= 50) {
            filledAt = candle.timestamp;
          }
        }
      }
    } else {
      // Bearish FVG - filled from below
      if (candle.high >= gapBottom) {
        const fillAmount = Math.min(candle.high - gapBottom, gapSize);
        const fillPercent = (fillAmount / gapSize) * 100;
        
        if (fillPercent > maxFill) {
          maxFill = fillPercent;
          if (fillPercent >= 50) {
            filledAt = candle.timestamp;
          }
        }
      }
    }
  }

  return {
    filled: maxFill >= 50,
    fillPercentage: maxFill,
    filledAt: filledAt
  };
}

/**
 * Check if price is in FVG zone
 */
export function isPriceInFVG(
  price: number,
  fvg: FairValueGap
): boolean {
  return price >= fvg.gapBottom && price <= fvg.gapTop;
}

/**
 * Get unfilled FVGs
 */
export function getUnfilledFVGs(fvgs: FairValueGap[]): FairValueGap[] {
  return fvgs.filter(f => !f.filled);
}

/**
 * Get partially filled FVGs (high probability zones)
 */
export function getPartiallyFilledFVGs(
  fvgs: FairValueGap[],
  minFill: number = 25,
  maxFill: number = 75
): FairValueGap[] {
  return fvgs.filter(f => 
    f.fillPercentage >= minFill && 
    f.fillPercentage <= maxFill
  );
}

/**
 * Find FVG near price level
 */
export function findFVGNearPrice(
  price: number,
  fvgs: FairValueGap[],
  tolerance: number = 0.1
): FairValueGap | null {
  const toleranceAmount = price * tolerance;

  for (const fvg of fvgs) {
    const midGap = (fvg.gapTop + fvg.gapBottom) / 2;
    if (Math.abs(price - midGap) <= toleranceAmount) {
      return fvg;
    }
  }

  return null;
}

/**
 * Calculate FVG strength based on size and volume
 */
export function calculateFVGStrength(
  fvg: FairValueGap,
  candles: Candle[],
  avgVolume: number
): number {
  const impulseCandle = candles[fvg.candleIndex];
  if (!impulseCandle) return 0;

  const gapSize = fvg.gapTop - fvg.gapBottom;
  const avgPrice = (fvg.gapTop + fvg.gapBottom) / 2;
  const gapPercent = (gapSize / avgPrice) * 100;

  // Volume factor (normalized)
  const volumeFactor = impulseCandle.volume / avgVolume;

  // Size factor (normalized to 0-1)
  const sizeFactor = Math.min(gapPercent / 1, 1); // 1% = max

  // Combined strength
  return Math.min((sizeFactor * 0.6 + volumeFactor * 0.4), 1);
}

/**
 * Get FVG confluence with other levels
 */
export function getFVGConfluence(
  fvg: FairValueGap,
  orderBlockZone: { high: number; low: number } | null,
  liquidityLevel: number | null
): number {
  let confluence = 0;

  // Check if FVG aligns with order block
  if (orderBlockZone) {
    const obMid = (orderBlockZone.high + orderBlockZone.low) / 2;
    const fvgMid = (fvg.gapTop + fvg.gapBottom) / 2;
    
    if (Math.abs(obMid - fvgMid) < fvg.gapTop - fvg.gapBottom) {
      confluence += 30;
    }
  }

  // Check if FVG aligns with liquidity level
  if (liquidityLevel) {
    if (liquidityLevel >= fvg.gapBottom && liquidityLevel <= fvg.gapTop) {
      confluence += 25;
    }
  }

  return confluence;
}
